src/sim/detail/InitializeTracks.cu

```diff
@@ -43,56 +43,54 @@ struct IsEqual
  * from either primary particles or secondaries. The new tracks are inserted
  * into empty slots (vacancies) in the track vector.
  */
-__global__ void init_tracks_kernel(const StatePointers            states,
-                                   const ParamPointers            params,
-                                   const TrackInitializerPointers inits,
-                                   size_type num_vacancies)
+__global__ void init_tracks_kernel(const ParamsDeviceRef         params,
+                                   const StateDeviceRef          states,
+                                   const TrackInitStateDeviceRef inits,
+                                   size_type                     num_vacancies)
 {
-    auto thread_id = KernelParamCalculator::thread_id().get();
-    if (thread_id < num_vacancies)
+    auto tid = KernelParamCalculator::thread_id();
+    if (!(tid < num_vacancies))
+        return;
+
+    // Get the track initializer from the back of the vector. Since new
+    // initializers are pushed to the back of the vector, these will be the
+    // most recently added and therefore the ones that still might have a
+    // parent they can copy the geometry state from.
+    const TrackInitializer& init
+        = inits.initializers[ThreadId(inits.initializers.size() - tid.get() - 1)];
+
+    // Thread ID of vacant track where the new track will be initialized
+    ThreadId vac_id(
+        inits.vacancies[ThreadId(inits.vacancies.size() - tid.get() - 1)]);
+
+    // Initialize the simulation state
     {
-        // Get the track initializer from the back of the vector. Since new
-        // initializers are pushed to the back of the vector, these will be the
-        // most recently added and therefore the ones that still might have a
-        // parent they can copy the geometry state from.
-        const TrackInitializer& init
-            = inits.initializers[inits.initializers.size() - thread_id - 1];
+        SimTrackView sim(states.sim, vac_id);
+        sim = init.sim;
+    }
 
-        // Index of the empty slot to create the new track in
-        ThreadId slot_id(
-            inits.vacancies[inits.vacancies.size() - thread_id - 1]);
+    // Initialize the particle physics data
+    {
+        ParticleTrackView particle(params.particles, states.particles, vac_id);
+        particle = init.particle;
+    }
 
-        // Initialize the simulation state
+    // Initialize the geometry
+    {
+        GeoTrackView geo(params.geometry, states.geometry, vac_id);
+        if (tid < inits.parents.size())
         {
-            SimTrackView sim(states.sim, slot_id);
-            sim = init.sim;
+            // Copy the geometry state from the parent for improved
+            // performance
+            ThreadId parent_id
+                = inits.parents[ThreadId(inits.parents.size() - tid.get() - 1)];
```

Comment:
Same here -- actually at this point I think you should define a
```c++
inline CELER_FUNCTION ThreadId from_end(size_type size, ThreadId cur_thread)
```
and reuse it in the three places (and the assertion can be a `CELER_EXPECT` at the top of the function)